test/global-rate-limit.test.js

```diff
@@ -347,3 +347,65 @@ test('does not override the preHandler as an array', t => {
     t.strictEqual(res.headers['x-ratelimit-remaining'], 1)
   })
 })
+
+test('variable max', t => {
+  t.plan(5)
+  const fastify = Fastify()
+  fastify.register(rateLimit, {
+    max: (req, key) => {
+      t.pass()
+      return +req.headers['secret-max']
+    },
+    timeWindow: 1000
+  })
+
+  fastify.get('/', (req, res) => { res.send('hello') })
+
+  fastify.inject({ url: '/', headers: { 'secret-max': 50 } }, (err, res) => {
+    t.error(err)
+    t.strictEqual(res.statusCode, 200)
+    t.strictEqual(res.headers['x-ratelimit-limit'], 50)
+    t.strictEqual(res.headers['x-ratelimit-remaining'], 49)
+  })
+})
+
+test('variable max contenders', t => {
+  t.plan(14)
+  const fastify = Fastify()
+  fastify.register(rateLimit, {
+    keyGenerator (req) { return req.headers['api-key'] },
+    max: (req, key) => { return key === 'pro' ? 3 : 2 },
+    timeWindow: 10000
+  })
+
+  fastify.get('/', (req, res) => { res.send('hello') })
+
+  fastify.inject({ url: '/', headers: { 'api-key': 'pro' } }, (err, res) => {
+    t.error(err)
+    t.strictEqual(res.statusCode, 200)
+    fastify.inject({ url: '/', headers: { 'api-key': 'pro' } }, (err, res) => {
```

Comment:
That is :boom: 
https://miro.medium.com/max/721/1*Co0gr64Uo5kSg89ukFD2dw.jpeg

Reply:
Yeah 😁 I wrote with async before Travis reminded me that doesn't work on node 6 😅